# <center> Preprocessing Files on CLI
<img src="Img/Youtube.jpg" width="300">

# 1. Introduction
## 1.1 What is Youtube? 
    
**YouTube** is a free video sharing website that makes it easy to watch online videos. You can even create and upload your own videos to share with others. Originally created in 2005, YouTube is now one of the most popular sites on the Web, with visitors watching around 2.6 billion users worldwide as of 2022. It's ranked as the second-most popular social network, and the only platform that has more active users than YouTube is Facebook.
    
## [1.2 Context](https://www.kaggle.com/datasets/datasnaek/youtube-new) 
    
YouTube (the world-famous video sharing website) maintains a list of the top trending videos on the platform. According to Variety magazine, "To determine the year's top-trending videos, YouTube uses a combination of factors including measuring users interactions (number of views, shares, comments and likes). Note that they're not the most-viewed videos overall for the calander year". Top performers on the YouTube trending list are music videos (such as the famous virile "Gangam Style"), celebrity and/or reality TV performances, and the random dude-with-a-camera viral videos that YouTube is well-known for. 
    
This dataset is a daily record of the top top trending YouTube videos. 
    
Note that this dataset is a structurally imporved version of [this dataset.](https://www.kaggle.com/datasnaek/youtube)
    
## [1.3 Content](https://www.kaggle.com/datasets/datasnaek/youtube-new)
    
This dataset includes several months (and counting) of data on daily trending YouTube videos. Data is included for the US, GB, DE, CA, and FR regions (USA, Great Britain, Germany, Canada, and France, respectively), with up to 200 listed trending videos per day. 
    
EDIT : Now includes data from RU, MX, KR, JP, and IN regions (Russia, Mexico, South Korea, Japan and India respectively) over the same time period. 
    
Each region's data is in a separate file. Data includes the video title, channel title, publish time, tags, views, likes and dislikes, description, and comment count. 
    
The data also includes a 'category_id' filed, which varies between regions. To retreive the categories for a specific video, find it in the associated JSON. One such file is included fore each of the five regions in the dataset. 
    
For more information on specific columns in the dataset refer to the [column metadata.](https://www.kaggle.com/datasnaek/youtube-new/data)
    
## 1.4 Features 
    
- video_id : Unique code of video 
- title : Title of video 	
- publishedAt : Published time of video 
- channelId : Unique code videoof youtube channel 
- channelTitle : Title of youtube channel
- categoryId : Unique code of category
- trending_date : Trending date of video 	
- tags : Tag of video
- view_count : Total view of video 
- likes : Counts of people like the video 
- dislikes : Counts of people dislike the video 
- comment_count : Counts of people comment on the video 
- thumbnail_link : Linkes of tunmbnails of the video 
- comments_disabled : Whether people can leave comments 
- ratings_disabled : Whether people can leave ratings 
- description : Description of video 
    
## 1.5 Purpose of this Project 
    
**Data Munging**, sometimes refereed to as **data munging**, is the process of transforming and mapping data from one "raw" data form into another format with the intent of making it more appropriate and valuable for a variety of downstream purposes such as analytics. The goal of data wrangling is to assure quality and useful data. Data analysts typically spend the majority of their time in the process of data wrangling compared to the actual analysis of the data. 
    
In this project, We will wrangle files in /Dataset merging files by continents.

# 2. Check Files 

## 2.1 Check Directory

In [1]:
!ls Dataset

BR_category_id.json	      IN_youtube_trending_data.csv
BR_youtube_trending_data.csv  JP_category_id.json
CA_category_id.json	      JP_youtube_trending_data.csv
CA_youtube_trending_data.csv  KR_category_id.json
DE_category_id.json	      KR_youtube_trending_data.csv
DE_youtube_trending_data.csv  MX_category_id.json
FR_category_id.json	      MX_youtube_trending_data.csv
FR_youtube_trending_data.csv  RU_category_id.json
GB_category_id.json	      RU_youtube_trending_data.csv
GB_youtube_trending_data.csv  US_category_id.json
IN_category_id.json	      US_youtube_trending_data.csv


In Dataset files, There are files with aliased coutnry name in front of its name. Each alisas stands for below : 

- BR : Brazil 
- CA : Canada
- DE : Germany 
- FR : France 
- GB : British 
- IN : India 
- JP : Japan
- KR : Korea 
- MX : Mexico 
- RU : Russia 
- US : USA 

We will merge countries by their continent. SA(South America : BR, MX), NA(North America : CA, US), EU(Europe : DE, FR, GB, RU), AS(Asia : IN, JP, KR). 

## 2.2 Check Files 

In [2]:
!head -n 2 Dataset/KR_youtube_trending_data.csv

Each file is consisted of one row of column and rows after. Columns are video_id, title, publishedAt, channelId, channelTitle, categoryId, trending_date, tags, view_count, likes, dislikes, comment_count, thumbnail_link, comments_disabled, ratings_disabled, description.

In [3]:
!wc Dataset/KR_youtube_trending_data.csv

   130555   8824365 155363335 Dataset/KR_youtube_trending_data.csv


KR_youtube_trending_data.csv file have 130555 rows and 155MB storage.

In [4]:
!grep -i '김종국' Dataset/KR_youtube_trending_data.csv | head -n 2



grep: 기록 오류: 파이프가 깨어짐


If we use grep command, we can find matching data more fastly using python and other programming languages. Using this command we can extract csv file with specific conditions, above example is grepping trending video with any char with '김종국'.

# 3. Preprocess with Commands

In [5]:
# Make files by continent 
!echo -n "" > Wrangled_Dataset/SA_youtube_trending_data.csv 
!touch Wrangled_Dataset/SA_youtube_trending_data.csv > Wrangled_Dataset/NA_youtube_trending_data.csv 
!touch Wrangled_Dataset/SA_youtube_trending_data.csv > Wrangled_Dataset/EU_youtube_trending_data.csv 
!touch Wrangled_Dataset/SA_youtube_trending_data.csv > Wrangled_Dataset/AS_youtube_trending_data.csv

In [6]:
# Add columns to files 
!head -n +1 Dataset/KR_youtube_trending_data.csv >> Wrangled_Dataset/SA_youtube_trending_data.csv 
!head -n +1 Dataset/KR_youtube_trending_data.csv >> Wrangled_Dataset/NA_youtube_trending_data.csv 
!head -n +1 Dataset/KR_youtube_trending_data.csv >> Wrangled_Dataset/EU_youtube_trending_data.csv 
!head -n +1 Dataset/KR_youtube_trending_data.csv >> Wrangled_Dataset/AS_youtube_trending_data.csv 

In [7]:
# Merge SA_youtube_trending_data.csv 
!tail -n +2 Dataset/BR_youtube_trending_data.csv >> Wrangled_Dataset/SA_youtube_trending_data.csv
!tail -n +2 Dataset/MX_youtube_trending_data.csv >> Wrangled_Dataset/SA_youtube_trending_data.csv 

In [8]:
# Merge NA_youtube_trending_data.csv 
!tail -n +2 Dataset/CA_youtube_trending_data.csv >> Wrangled_Dataset/NA_youtube_trending_data.csv
!tail -n +2 Dataset/US_youtube_trending_data.csv >> Wrangled_Dataset/NA_youtube_trending_data.csv 

In [9]:
# Merge EU_youtube_trending_data.csv 
!tail -n +2 Dataset/GB_youtube_trending_data.csv >> Wrangled_Dataset/EU_youtube_trending_data.csv
!tail -n +2 Dataset/DE_youtube_trending_data.csv >> Wrangled_Dataset/EU_youtube_trending_data.csv 
!tail -n +2 Dataset/FR_youtube_trending_data.csv >> Wrangled_Dataset/EU_youtube_trending_data.csv
!tail -n +2 Dataset/RU_youtube_trending_data.csv >> Wrangled_Dataset/EU_youtube_trending_data.csv 

In [10]:
# Merge AS_youtube_trending_data.csv 
!tail -n +2 Dataset/KR_youtube_trending_data.csv >> Wrangled_Dataset/AS_youtube_trending_data.csv
!tail -n +2 Dataset/IN_youtube_trending_data.csv >> Wrangled_Dataset/AS_youtube_trending_data.csv 
!tail -n +2 Dataset/JP_youtube_trending_data.csv >> Wrangled_Dataset/AS_youtube_trending_data.csv

In [11]:
# Check the result 
!wc -l Dataset/KR_youtube_trending_data.csv
!wc -l Dataset/IN_youtube_trending_data.csv 
!wc -l Dataset/JP_youtube_trending_data.csv 

# Total result 
!wc -l Wrangled_Dataset/AS_youtube_trending_data.csv

130555 Dataset/KR_youtube_trending_data.csv
127528 Dataset/IN_youtube_trending_data.csv
133588 Dataset/JP_youtube_trending_data.csv
391669 Wrangled_Dataset/AS_youtube_trending_data.csv


# 4. Preprocess with CSVkits

In [12]:
# Merge SA_youtube_trending_data.csv 
!csvstack -n country -g "Brazil,Mexico" Dataset/BR_youtube_trending_data.csv Dataset/MX_youtube_trending_data.csv > Wrangled_Dataset/SA_youtube_trending_data2.csv 

# Merge NA_youtube_trending_data.csv 
!csvstack -n country -g "Canada,USA" Dataset/CA_youtube_trending_data.csv Dataset/US_youtube_trending_data.csv > Wrangled_Dataset/NA_youtube_trending_data2.csv 

# Merge EU_youtube_trending_data.csv 
!csvstack -n country -g "Britsh,Germany,France,Russia" Dataset/GB_youtube_trending_data.csv Dataset/DE_youtube_trending_data.csv Dataset/FR_youtube_trending_data.csv Dataset/RU_youtube_trending_data.csv > Wrangled_Dataset/EU_youtube_trending_data2.csv 

# Merge AS_youtube_trending_data.csv 
!csvstack -n country -g "Korea,India,Japan" Dataset/KR_youtube_trending_data.csv Dataset/JP_youtube_trending_data.csv Dataset/IN_youtube_trending_data.csv > Wrangled_Dataset/AS_youtube_trending_data2.csv 

Error: line contains NUL
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='Dataset/FR_youtube_trending_data.csv' mode='rt' encoding='utf-8-sig'>
Error: line contains NUL
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='Dataset/JP_youtube_trending_data.csv' mode='rt' encoding='utf-8-sig'>


In [13]:
# Check the result 
!wc -l Dataset/KR_youtube_trending_data.csv
!wc -l Dataset/IN_youtube_trending_data.csv 
!wc -l Dataset/JP_youtube_trending_data.csv 

# Total result 
!wc -l Wrangled_Dataset/AS_youtube_trending_data2.csv

130555 Dataset/KR_youtube_trending_data.csv
127528 Dataset/IN_youtube_trending_data.csv
133588 Dataset/JP_youtube_trending_data.csv
303428 Wrangled_Dataset/AS_youtube_trending_data2.csv
